src/Processors/Formats/Impl/NpyOutputFormat.cpp

```diff
@@ -0,0 +1,243 @@
+    {
+        case TypeIndex::Int8: numpy_data_type = NumpyDataType('<', 'i', sizeof(Int8)); break;
+        case TypeIndex::Int16: numpy_data_type = NumpyDataType('<', 'i', sizeof(Int16)); break;
+        case TypeIndex::Int32: numpy_data_type = NumpyDataType('<', 'i', sizeof(Int32)); break;
+        case TypeIndex::Int64: numpy_data_type = NumpyDataType('<', 'i', sizeof(Int64)); break;
+        case TypeIndex::UInt8: numpy_data_type = NumpyDataType('<', 'u', sizeof(UInt8)); break;
+        case TypeIndex::UInt16: numpy_data_type = NumpyDataType('<', 'u', sizeof(UInt16)); break;
+        case TypeIndex::UInt32: numpy_data_type = NumpyDataType('<', 'u', sizeof(UInt32)); break;
+        case TypeIndex::UInt64: numpy_data_type = NumpyDataType('<', 'u', sizeof(UInt64)); break;
+        case TypeIndex::Float32: numpy_data_type = NumpyDataType('<', 'f', sizeof(Float32)); break;
+        case TypeIndex::Float64: numpy_data_type = NumpyDataType('<', 'f', sizeof(Float64)); break;
+        case TypeIndex::FixedString: numpy_data_type = NumpyDataType('|', 'S', assert_cast<const DataTypeFixedString *>(type.get())->getN()); break;
+        case TypeIndex::String: numpy_data_type = NumpyDataType('|', 'S', 0); break;
+        case TypeIndex::Array: return getNumpyDataType(assert_cast<const DataTypeArray *>(type.get())->getNestedType());
+        default: nested_data_type = type; return false;
+    }
+
+    nested_data_type = type;
+    return true;
+}
+
+void NpyOutputFormat::consume(Chunk chunk)
+{
+    if (!invalid_shape)
+    {
+        num_rows += chunk.getNumRows();
+        const auto & column = chunk.getColumns()[0];
+
+        if (!is_initialized)
+        {
+            initShape(column);
+            is_initialized = true;
+        }
+
+        if (!checkShape(column))
+        {
+            invalid_shape = true;
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "ClickHouse doesn't support object types, cannot format ragged nested sequences (which is a list of arrays with different shapes)");
+        }
+    }
+}
+
+void NpyOutputFormat::initShape(const ColumnPtr & column)
+{
+    auto type = data_type;
+    ColumnPtr nested_column = column;
+    while (type->getTypeId() == TypeIndex::Array)
+    {
+        const auto * array_column = assert_cast<const ColumnArray *>(nested_column.get());
+
+        numpy_shape.push_back(array_column->getOffsets()[0]);
+
+        type = assert_cast<const DataTypeArray *>(type.get())->getNestedType();
+        nested_column = array_column->getDataPtr();
+    }
+}
+
+bool NpyOutputFormat::checkShape(const ColumnPtr & column)
+{
+    auto type = data_type;
+    ColumnPtr nested_column = column;
+    int dim = 0;
+    while (type->getTypeId() == TypeIndex::Array)
+    {
+        const auto * array_column = assert_cast<const ColumnArray *>(nested_column.get());
+        const auto & array_offset = array_column->getOffsets();
+
+        for (size_t i = 1; i < array_offset.size(); ++i)
```

Comment:
```suggestion
        for (size_t i = 0; i < array_offset.size(); ++i)
```

The size of i-th array is `offsets[i] - offsets[i - 1]`, so for the first array size will be `offsets[0] - offsets[-1]`. And it's ok to use `offsets[-1]` as we use our own vector implementation and `offsets[-1]` is always `0`